scripts/version_buckets.py

```diff
@@ -0,0 +1,53 @@
+"""
+Reads all the current buckets, ensures that versioning is switched on for those
+that don't yet have it and sets the life cycle configuration to send
+non-current versions of files to glacier storage after 30 days.
+
+This script should only be run once, when the following PR lands in master in
+the control panel:
+
+https://github.com/ministryofjustice/analytics-platform-control-panel/pull/816
+
+(From that moment on, all S3 buckets will be configured in this way by the
+control panel, as a default.)
+"""
+import boto3
+
+
+s3 = boto3.client('s3')
+buckets = s3.list_buckets()
+for bucket_dict in buckets["Buckets"]:
+    bucket_name = bucket_dict["Name"]
+    if not bucket_name.startswith("alpha"):
+        # Only deal with buckets created and used by/for the control panel
+        # (always prefixed with "alpha").
+        continue
+    print(f"Working on {bucket_name}.")
+    bucket = boto3.resource("s3").Bucket(bucket_name)
+    # Add versioning if not already set.
+    versioning = bucket.Versioning()
+    if not versioning.status == "Enabled":
+        print(f"Enabling versioning for {bucket_name}.")
+        versioning.enable()
+    # Set life cycle rule to send non-current versions of files to glacier
+    # storage after 30 days.
+    lifecycle_id = f"{bucket_name}_lifecycle_configuration"
+    print(f"Setting lifecycle {lifecycle_id} for bucket {bucket_name}.")
+    lifecycle_conf = boto3.client("s3").put_bucket_lifecycle_configuration(
+        Bucket=bucket_name,
+        LifecycleConfiguration={
+            "Rules": [
+                {
+                    "ID": lifecycle_id,
+                    "Status": "Enabled",
+                    "Prefix": "",
+                    "NoncurrentVersionTransitions": [
+                        {
+                            'NoncurrentDays': 30,
+                            'StorageClass': 'GLACIER',
+                        },
+                    ]
+                },
+            ]
+        }
+    )
```

Comment:
Again this script is effecting a large number of S3 buckets, my main concern with this is that it always seem to set (put) the lifecycle regardless of whether the versioning was enabled or not.

For example there are S3 buckets created with terraform (or similar) and they may have a different lifecycle policy and this would change them conflicting with whatever terraform (or similar) wants.

Maybe a safer thing to do would be to only set this lifecycle policy if the versioning is not enabled, at least you reduce the chance to change this on buckets not created by the CP before your addition?

Reply:
Yeah, it makes me nervous running such a big job like.How about the script also checks the existing versioning setting and any lifecycle policies. If we run it '--dry-run' we can review the report before doing any changes. And maybe it's overkill, but maybe we need to do a gradual roll out, say 10%, give it a couple of days, then 25%, 50%, 100%, just in case?

Reply:
Ack.

I think a `--dry-run` flag is good so we can be sure we're identifying the right buckets and transforming them in the correct way. However, once we're happy with the planned changes, I don't see why we can't just do them all at once...? I'd welcome thoughts / advice on this (am I missing something here..?)

Reply:
I'd also move the `put_bucket_lifecycle_configuration()` inside the `if not versioning.status == "Enabled":` to reduce the impact of this, it's kinda of a migration and no point in overwriting this on bucket which already have versioning enabled as they may be have been created by terraform, not Control Panel.

Reply:
> I don't see why we can't just do them all at once...?

Because system health is really important to us. We have to make a judgement on what defensive deployment techniques are appropriate or overkill.

To compare, when we deploy software with helm, it does blue-green deploys and we can rollback reliably with one command. When we deploy infrastructure with terraform, the state before and after are stored in git, and again it has the tools to easily and reliably roll-back. But this PR's migration by script is rather more blunt. So whilst there's nothing specific about this change that looks risky, I think some caution is necessary. The risk would be lower if the team had done recent changes of this type.

Reply:
Ack.

So what approach to running this script would be most appropriate. For instance I can:

* Grab the names of all the buckets to be changed (from the PoV of the script).
* We confirm these look right (eyeball mk.1)
* We group these buckets and run the script against each bucket at a time.
* After migrating each bucket, and before migrating the next bucket, we confirm (eyeball mk.1) that the migration is as we expect.
* I suggest early groups of buckets are small and contain buckets that belong to us.

Does this help..?

Reply:
👍 Sounds good. 

How about running the script `--dry-run` and pasting the output in this PR, so we can review it's intended actions? It can show for each bucket if it already has versioning and lifecycle config, and it wouldn't need to do any changes.

In terms of doing only some buckets, you could perhaps add a `-n` to do only a limited number of them. My feeling is we can simply sort them in reverse alphabetical order, to leave the covid and nomis ones out of the first batches. Try `-n 1` then `-n 10`, and if there are no shouts of pain after couple of days, do `-n 50`, or something.